vmm/src/api/http/http_endpoint.rs

```diff
@@ -196,55 +209,118 @@ vm_action_put_handler_body!(VmSendMigration);
-// number provided by the client is almost certainly invalid within the
-// hypervisor process.
-//
-// To address this, file descriptors are transmitted over UNIX domain sockets
-// using the SCM_RIGHTS control message mechanism. The Linux kernel handles
-// these messages by duplicating the referenced FDs into the receiving process,
-// thereby ensuring they are valid and usable in the target context.
-//
-// Once these valid file descriptors are received here, we integrate the actual
-// FDs into the virtual machine configuration, allowing the virtio-net device to
-// function correctly with its backing network resources.
+/// Applies FDs to the network config of a given device, as part of the special
+/// handling for virtio-net devices backed by network FDs.
+///
+/// # Prerequisites
+/// This mechanism requires that the control software connects to Cloud
+/// Hypervisor uses a UNIX domain socket and that it passes file descriptors
+/// (FDs) via ancillary messages—specifically using the SCM_RIGHTS mechanism
+/// described in cmsg(3). These ancillary messages must accompany the primary
+/// payload (HTTP JSON REST API in this case).
+///
+/// # Motivation and Rationale
+/// File descriptor numbers are specific to the process in which they are created
+/// and therefore cannot be assumed to have any meaning in another process. When
+/// configuring virtio-net devices via Cloud Hypervisor’s HTTP REST API, any FD
+/// number provided by the client is almost certainly invalid within the
+/// hypervisor process.
+///
+/// To address this, file descriptors are transmitted over UNIX domain sockets
+/// using the SCM_RIGHTS control message mechanism. The Linux kernel handles
+/// these messages by duplicating the referenced FDs into the receiving process,
+/// thereby ensuring they are valid and usable in the target context.
+///
+/// Once these valid file descriptors are received here, we integrate the actual
+/// FDs into the virtual machine configuration, allowing the virtio-net device to
+/// function correctly with its backing network resources.
+fn apply_new_fds_to_cfg<T>(
+    // List of new files (well, actually FDs) that back up a virtio-net device.
+    files: Vec<File>,
+    // List of network configurations where each network can have `n` FDs.
+    network_cfgs: &mut [&mut T],
+    // Callback to return the ID.
+    network_cfg_extract_id: &impl Fn(&T) -> Option<&str>,
+    // Callback to extract the amount of expected FDs.
+    network_cfg_extract_num_fds_fn: &impl Fn(&T) -> usize,
+    // Callback to extract the FDs that are part of the type (transmitted via
+    // the HTTP body)
+    network_cfg_extract_fds_fn: &impl Fn(&T) -> Option<&[i32]>,
+    // Callback to set any FDs in the type to the new value. The new value
+    // is either `Some` with a non-empty Vector or `None`.
+    network_cfg_replace_fds: &impl Fn(&mut T, Option<Vec<i32>>),
+) -> Result<(), HttpError> {
```

Comment:
Figuring out how to call this function has an enormous cognitive complexity. Is there really no better way to do it?